rllab/tf/algos/ddpg.py

```diff
@@ -0,0 +1,319 @@
+                tc.layers.l2_regularizer(self.actor_weight_decay),
+                weights_list=self.actor.regularizable_vars)
+            action_loss += actor_reg
+        actor_train_op = self.optimizer(self.actor_lr).minimize(
+            action_loss, var_list=self.actor.trainable_vars)
+
+        f_train_actor = tensor_utils.compile_function(
+            inputs=[states], outputs=[actor_train_op, action_loss])
+
+        self.opt_info = dict(
+            f_train_actor=f_train_actor,
+            f_train_critic=f_train_critic,
+            replay_buffer=replay_buffer,
+            target_critic=target_critic,
+            target_actor=target_actor)
+
+    def _learn(self):
+        """
+        Perform algorithm optimizing.
+
+        Returns:
+            action_loss: Loss of action predicted by the policy network.
+            qval_loss: Loss of q value predicted by the q network.
+            ys: y_s.
+            qval: Q value predicted by the q network.
+        """
+        replay_buffer = self.opt_info["replay_buffer"]
+        target_actor = self.opt_info["target_actor"]
+        target_critic = self.opt_info["target_critic"]
+        f_train_critic = self.opt_info["f_train_critic"]
+        f_train_actor = self.opt_info["f_train_actor"]
+
+        transitions = replay_buffer.random_sample(self.batch_size)
+        observations = transitions["observations"]
+        rewards = transitions["rewards"]
+        actions = transitions["actions"]
+        terminals = transitions["terminals"]
+        next_observations = transitions["next_observations"]
+        rewards = rewards.reshape(-1, 1)
+        terminals = terminals.reshape(-1, 1)
+
+        next_actions = target_actor.get_actions(next_observations)
+        next_qvals = target_critic.get_qval(next_observations, next_actions)
+
+        ys = rewards + (1. - terminals) * self.discount * next_qvals
+
+        _, action_loss = f_train_actor(observations)
+        _, qval_loss, qval = f_train_critic(ys, observations, actions)
+
+        return action_loss, qval_loss, ys, qval
+
+    def _predict(self, state):
+        """
+        Select action a_t = mu(s_t) + N according to the input state and
+        exploration strategy.
+
+        Args:
+            state: Current env observation.
+
+        Returns:
+            action: Sampled action.
+        """
+        action = self.actor.get_action(np.array(state))
+        noise = self.es.get()
+        action = action + noise
+        action = np.clip(action, self.action_range[0], self.action_range[1])
+        return action
+
+    @overrides
+    def train(self, sess=None):
+        self._initialize()
+
+        replay_buffer = self.opt_info["replay_buffer"]
+        created_session = True if (sess is None) else False
+        if sess is None:
+            sess = tf.Session()
+            sess.__enter__()
+
+        sess.run(tf.global_variables_initializer())
+
+        state = self.env.reset()
```

Comment:
To conform with the naming convention of other algos, it would be better to name this as obs.